code/sys/sys_main.c

```diff
@@ -642,6 +646,81 @@ void Sys_ParseArgs( int argc, char **argv )
 	}
 }
 
+/*
+=================
+Sys_ParseProtocolUri
+
+This parses a protocol URI, e.g. "quake3://connect/example.com:27950"
+to a string that can be run in the console, or a null pointer if the
+operation is invalid or unsupported.
+At the moment only the "connect" command is supported.
+=================
+*/
+#ifdef PROTOCOL_HANDLER
+char *Sys_ParseProtocolUri( char *uri )
+{
+	// Both "quake3://" and "quake3:" can be used
+	char *command;
+	if (strstr(uri, PROTOCOL_HANDLER "://") == uri)
+	{
+		command = uri + strlen(PROTOCOL_HANDLER "://");
+	}
+	else if (strstr(uri, PROTOCOL_HANDLER ":") == uri)
+	{
+		command = uri + strlen(PROTOCOL_HANDLER ":");
+	}
+	else
+	{
+		Com_Printf ("Sys_ParseProtocolUri: unsupported protocol.\n");
+		return NULL;
+	}
+	if (command[0] == '\0') {
+		return NULL;
+	}
+	Com_Printf ("Sys_ParseProtocolUri: %s\n", command);
+
+	// At the moment, only "connect/hostname:port" is supported
+	// For safety reasons, the "hostname:port" part can only
+	// contain characters from [a-zA-Z0-9.:-]
+	if (strstr(command, "connect/") != command)
+	{
+		Com_Printf ("Sys_ParseProtocolUri: unsupported command.\n");
+		return NULL;
+	}
+	char *arg = strchr(command, '/');
+	if (arg == NULL || arg[1] == '\0' || arg[1] == '?')
+	{
+		Com_Printf ("Sys_ParseProtocolUri: missing argument.\n");
+		return NULL;
+	}
+	*arg = ' ';
+	arg++;
+
+	// Check for any unsupported characters
+	for (int i=0; arg[i] != '\0'; i++)
+	{
+		if (arg[i] == '?') {
+			// For forwards compatibility, any query string parameters are ignored (e.g. "?password=abcd")
+			// However, these are not passed on macOS, so it may be a bad idea to add them.
+			arg[i] = '\0';
+			break;
+		}
+
+		if (isalpha(arg[i]) == 0 && isdigit(arg[i]) == 0
+		 && arg[i] != '.' && arg[i] != ':' && arg[i] != '-')
+		{
+			Com_Printf ("Sys_ParseProtocolUri: hostname contains unsupported characters.\n");
+			return NULL;
+		}
+	}
+
+	int bufsize = strlen( command ) + 1;
+	char *data = Z_Malloc( bufsize );
+	Q_strncpyz( data, command, bufsize );
+	return data;
```

Comment:
Untested, but should probably be something like:
```suggestion
	return CopyString(command);
```

Reply:
Thanks, will test this.

Reply:
Seems to work. 👍